### DISCUSSION TITLE
How to do 'only' forecast without make_evaluation_predictions or setting up the test dataset?

### DISCUSSION THREAD

**Discussion by Msaleh87** [awslabs/gluonts#2350] [Q&A]
So I do the forecast and compare it to the test dataset as follows: 

**First I set up the dataset for testing:**

```
test_data = ListDataset(
    [
        {"start": df_input.index[0], "target": df_input.Load[:'2022-03-13 04:50:00'], 
         'feat_dynamic_real': [df_input.DofW[:'2022-03-13 04:50:00']], 
        'time_feat': [time_feat_df.DayofWeek[:'2022-03-13 04:50:00'],
                     time_feat_df.Hour[:'2022-03-13 04:50:00'],]}, # one day
        {"start": df_input.index[0], "target": df_input.Load[:'2022-03-14 04:50:00'], 
         'feat_dynamic_real': [df_input.DofW[:'2022-03-14 04:50:00']],
        'time_feat': [time_feat_df.DayofWeek[:'2022-03-14 04:50:00'],
                    time_feat_df.Hour[:'2022-03-14 04:50:00'],]}   # two days
    
    ],

freq = "5min")
```
**Then I run make_evaluation_predictions :** 

```
from gluonts.evaluation.backtest import make_evaluation_predictions
forecast_it, ts_it = make_evaluation_predictions(
    dataset = test_data,
    predictor = predictor,
    num_samples = 10
)
forecasts = list(forecast_it)
tss = list(ts_it)
```

Now, I want to, **only**, predict the next horizon without using the test dataset. How to do so? I explored the documentation, but the documentation just shows how to evaluate your model against the test dataset. Still, it doesn't show you how to only forecast for a specified future horizon. I am asking about this because it would help in online operations. 
I am assuming there is another function besides make_evaluation_predictions to forecast, where I specify the model/predictor that I have trained already and the future horizon and it would provide me with the forecast. Or should I just use the make_evaluation_predictions with a dummy dataset and extract the forecasted data. Thanks

**Comment by lostella** [comment]
@Msaleh87  you do this with

```python
forecasts = predictor.predict(data)
```

And then iterate over `forecasts` (or just turn it into a `list` if the dataset is small enough).

**Reply by Msaleh87** [reply]
Hi @lostella, thank you for getting back to me, which data in the following command? `forecasts = predictor.predict(data)`

So what I am trying to do is to predict the electric load for the next 48 hours (576 future data points). I just want to extract the 576 future data points right after the end/ last data point of the training data. I have trained the model and I have my predictor. I tried the command you mentioned as follows:

```
forecasts = predictor.predict(training_data) # Does it use the prediction_length provided in the estimator?
forecasted_data = list(forecasts)   
```
I got the following error:

![Error](https://user-images.githubusercontent.com/42920113/194900893-cd14bdb0-456f-457b-89af-b7b092c3258d.png)

Also, I didn't get "iterate over `forecasts`." I tried forecasted_data = next(iter(forecasts)) and I got the same error
I was hoping to have the forecast and then use forecasted_data.mean to extract the data like we do with the make_evaluation_predictions function.

**Reply by Msaleh87** [reply]
So in the `make_evaluation_predictions` function, you provide the test dataset, the model/ predictor you trained, and the num_samples. I was hoping to have the forecast only (forecast_it), and if there is a function that I give it the model/ predictor and num_samples without the dataset and it predicts the future horizon (576 points/ 48 hrs ahead)

```
forecast_it, ts_it = make_evaluation_predictions(
    dataset = test_data,
    predictor = predictor,
    num_samples = 20
)
```

**Reply by Msaleh87** [reply]
The reason I am asking, I have a controller for a small electric grid (a microgrid) that needs the forecast of the next 48-hrs (576 data points) of the electric load every time step (every 5 mins). So I have my model trained already, how to use it to provide the future/ next forecast horizon without the test dataset.

**Reply by Msaleh87** [reply]
> ```python
> forecasts = predictor.predict(data)
> ```

@lostella Still stuck, is data for the whole historical data? then I should do list(forecasts) if the data is small, or iter! I tried that:

```
forecasts_DDAARR = predictor.predict(dataset = df_input, prediction_length = 576, lead_time = 0,num_samples = 20) 
#list(forecasts_DDAARR)
next(iter(forecasts_DDAARR)) # the error is reffering to this line <----
```

I got the following error:
![Screenshot 2022-12-23 173017](https://user-images.githubusercontent.com/42920113/209410700-5128995a-cc17-45de-b014-a31ec2a96b2a.png)

**Reply by sysublackstar** [reply]
I met the same error when use `predictor.predict()`🥲 and the index 0 size is exactly the same as index 1 size + prediction_length like yours.
Have you solved the problem?

**Reply by Msaleh87** [reply]
Hi @sysublackstar, yah I was able to slove the problem, check this link: https://github.com/awslabs/gluonts/discussions/2520

**Reply by sysublackstar** [reply]
Thanks for your issue!! I know what's wrong...the dataset should contain `feat_dynamic_real` about future🤣 @Msaleh87

**Comment by Msaleh87** [comment]
Hi @lostella, so I tried the following: `forecasts = predictor.predict(dataset = training_data, prediction_length = 576, lead_time = 0,num_samples = 20) ` which worked, but then I am stuck, I don't know how to extract the forecasted data. I checked the [source code for gluonts.model.predictor](https://ts.gluon.ai/stable/_modules/gluonts/model/predictor.html#Predictor) but still wasn't able to go further.
In the documentation, it was mentioned the following: **"A Predictor defines the predictor.predict method of a given predictor. This method takes the test dataset, it passes it through the prediction network to take the predictions, and yields the predictions. You can think of the Predictor object as a wrapper of the prediction network that defines its predict method.**" Which means it takes the test dataset. I am looking for a method/ function that uses the trained model/ predictor and the training dataset and provide me with the future forecast only. Is this method/ function exist? Or should I extract the forecasted data using the `make_evaluation_predictions` and `forecast.mean` and use a dummy test dataset to make them work?

My dataset is just 2 columns, time and electric load values as shown in the following pic:
![Training dataset](https://user-images.githubusercontent.com/42920113/195144318-cbebb0a3-12d3-4cb6-a13d-9a9d6cd3b18d.png)

How to extract the forecasted data? And am I supposed to provide the `dataset = training_data` or should I use something else?

Thanks!

**Comment by Msaleh87** [comment]
So here is my solution, it might be tedious, but it kind of works. First, create a dummy test dataset starting with the timestamp of the current time:

```
from datetime import datetime, timedelta
now = datetime.today() # will give you the current timestamp
Future_timestamp = now + timedelta(hours=48)

# If you want to insert it manually for testing: now = datetime(2022, 3, 13, 4, 54, 0)
Dummy = [0] * 480      # dummy list full of zeros # 2 days * 24 hrs * (60/6) time steps = 480. This number depends on your data resolution

df = pd.DataFrame({'Time':pd.date_range(start = now, periods=480, freq = '6min'),  
                  'Load': Dummy}) # make sure to name the column similar to what you have in your training dataset
df = df.set_index('Time')

New_df = pd.concat([df_input[:] , df],axis = 0)         # df_input[:] is your training dataset

test_data_dummy = ListDataset(
    [
        {"start": New_df.index[0], "target": New_df.Load[:], 
         'feat_dynamic_real': [df_input.DofW[:New_df.index[-1:][0]]], 
        'time_feat': [time_feat_df.DayofWeek[:New_df.index[-1:][0]],
                     time_feat_df.Hour[:New_df.index[-1:][0]],]}
    # New_df.index[-1:][0]] is the last time stamp in the new created dataset. or you could use Future_timestamp
    ],

freq = "6min")
```

Then run the normal code using `make_evaluation_predictions`:

```
forecast_it, ts_it = make_evaluation_predictions(
    dataset = test_data_dummy,
    predictor = predictor,
    num_samples = 20
)
forecasts = list(forecast_it)
tss = list(ts_it) # this one is not needed, just for checking
```

Then for extracting and plotting: 
```
ts_e = tss[0]
forecast_e = forecasts[0]

f_length = len(forecast_e.mean)

data = {
        'timestamp': ts_e.index[-f_length:].strftime("%Y-%m-%d %H:%M:%S"),
        'forecast': forecast_e.mean,
        'Real data' : ts_e[-f_length:][0].values
    }
dff = pd.DataFrame(data)
dff = dff.set_index('timestamp')
dff.plot()
```
Of course with this, I'll need to update the training dataset (which is in this case df_input) with every new reading from the meter that comes in, but this would be the general idea.

I tried using the training dataset right away as follows and I got the following error: `forecasts = predictor.predict(dataset = df_input[:'2022-03-07 04:50:00'], prediction_length = 480,num_samples = 20)` 

I know I might be doing something wrong, or might be a simple mistake, but trying to brainstorm.

![Error2](https://user-images.githubusercontent.com/42920113/195408454-0190f75e-a148-466e-8a17-2eae97475e33.png)
